resources/tests/org.eclipse.core.tests.resources/plugin.xml

```diff
@@ -534,18 +534,33 @@
       <content-type id="org.eclipse.core.tests.resources.nature_associated_2"/>       
 </extension>
 
+<!-- ORDER MATTER HERE START -->
+<!-- We want a particular oder of the providers and even if the first one is declared first it should be the last in the chain (and thus never be asked!) -->
+<extension
+      point="org.eclipse.core.resources.refreshProviders">
+   <refreshProvider
+         class="org.eclipse.core.tests.resources.refresh.LowerPriorityProvider"
+         name="org.eclipse.core.tests.resources.testRefreshProvider"
+         priority="1"
+         />
+</extension>
 <extension
       point="org.eclipse.core.resources.refreshProviders">
    <refreshProvider
          class="org.eclipse.core.tests.resources.refresh.TestRefreshProvider"
-         name="org.eclipse.core.tests.resources.testRefreshProvider"/>
+         name="org.eclipse.core.tests.resources.testRefreshProvider"
+         priority="5"
+         />
 </extension>
 <extension
       point="org.eclipse.core.resources.refreshProviders">
    <refreshProvider
          class="org.eclipse.core.tests.resources.refresh.Bug316182RefreshProvider"
-         name="org.eclipse.core.tests.resources.bug316182RefreshProvider"/>
+         name="org.eclipse.core.tests.resources.bug316182RefreshProvider"
+         priority="10"
```

Comment:
> Shouldn't one create new "supplemental" attribute for additional optional providers that could be contributed?

Thinking a bit longer, this also doesn't make sense. Either the provider can refresh or not, two or more of different providers on a project wouldn't make things better and just do same work multiple times.

What you actually want in https://github.com/eclipse-platform/eclipse.platform/issues/299 is a possibility for 3rd party code to **selectively** switch auto-refresh on/off for some specific (m2e) projects.

Reply:
> if the "real" one will win during tests on Windows?

Only if the windows fragment is included in the test (I talked with @HannesWell that he want to test this as well under "real" Windows)

> Missing explanation in extension point what happens if two providers have same priority.

That is true, I just wondering how far one want to go here, I don not expect many providers are implemented, actually we only have one for windows ...

> Should we disallow that & throw an exception if there is a second provider with same prio defined?
> If we don't do that, "random" one will be selected, based on extension point reading order by the OSGI.

Actually this could be a valid case, if we have one provider that only supports local resource and one that only supports remote ones, still I think we will most probably define a total order here.

Current behavior is to just call *all* providers... that seems not what one wants but again, as there is only one in reality...

> Also looking on the RefreshProviderTest, I wonder how that can work if we want both test refresh provider and real one (like Win32RefreshProvider) to be tested.

Real testing is actually not possible .. the tests mostly check that some methods are called, and one provider is even misused to check a completely different thing :-\

> For me it looks like original extension point was implemented only with single production implementation in mind, so it assumed there will be no multiple "production" refresh providers contributed. Now with introducing priority we actually create expectation that it is allowed to contribute multiple but at the end only the "highest" will be used.

probably, at laest the code is not very well prepared currently... but having a priority and using the highest that can handle the resource seems mostly resemble the "only one provider" mindset?!?

> What you actually want in https://github.com/eclipse-platform/eclipse.platform/issues/299 is a possibility for 3rd party code to selectively switch auto-refresh on/off for some specific (m2e) projects.

Actually I want that no one has to think about it or configure anything :-)
But the API is somehow prepared for such a case, but in the moment it is only all on/off ..